src/aproxy.py

```diff
@@ -123,85 +277,73 @@ def _is_proxy_reachable(self, host: str, port: int = DEFAULT_PROXY_PORT) -> bool
             logger.error("Proxy %s:%s is not reachable: %s", host, port, e)
             return False
 
     # ---------------- nftables ----------------
 
     def _get_primary_ip(self) -> str:
-        """Detect machine's primary IP for DNAT target."""
-        try:
-            with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
-                s.connect(("8.8.8.8", 80))
-                return s.getsockname()[0]
-        except (socket.timeout, ConnectionRefusedError, OSError) as e:
-            logger.warning("Falling back to 127.0.0.1 for DNAT target: %s", e)
-            return "127.0.0.1"
-
-    def _resolve_hostname_to_ips(self, hostname: str) -> list[str]:
-        """Resolve a hostname to its corresponding IP addresses.
-
-        Args:
-            hostname: The hostname to resolve.
+        """Get this unit's primary IP using Juju binding.
 
         Returns:
-            A list of resolved IP addresses.
+            The unit's bound private IP address.
+
+        Raises:
+            TopologyUnavailableError: If the binding or network information is unavailable.
         """
-        try:
-            return list(
-                {
-                    str(info[4][0])
-                    for info in socket.getaddrinfo(hostname, None, proto=socket.IPPROTO_TCP)
-                }
+        relation = self.charm.model.get_relation(RELATION_NAME)
+        binding = self.charm.model.get_binding(RELATION_NAME)
+
+        if not relation or not binding:
+            raise TopologyUnavailableError(
+                f"Relation '{RELATION_NAME}' binding not available when trying to get topology."
             )
-        except socket.gaierror as e:
-            logger.error("Failed to resolve hostname %s: %s", hostname, e)
-            return []
 
-    def _get_no_proxy_ips(self) -> list[str]:
-        """Convert no-proxy entries to a list of IP addresses.
+        if binding.network is None:
+            raise TopologyUnavailableError(
+                f"Relation '{RELATION_NAME}' network not available when trying to get unit IP."
+            )
 
-        Returns:
-            A list of resolved IP addresses.
-        """
-        ip_list: list[str] = []
-        for entry in map(str.strip, self.config.no_proxy.split(",")):
-            if not entry:
-                continue
-            try:
-                # Check if entry is a valid IP or CIDR
-                ipaddress.ip_network(entry, strict=False)
-                ip_list.append(entry)
-            except ValueError:
-                # Not an IP, attempt to resolve as hostname
-                resolved_ips = self._resolve_hostname_to_ips(entry)
-                ip_list.extend(resolved_ips)
-        return ip_list
-
-    def _render_nftables_rules(self) -> str:
+        units_ip: list[str] = [
+            unit_data.get("private-address", "")
+            for _, unit_data in relation.data.items()
+            if unit_data.get("private-address", "")
+        ]
```

Comment:
Enclose this code checking the log verbosity with `getEffectiveLevel` to avoid executing this code if we're not in debug mode